chain4travel/camino-wallet
# fix(erc721): display custom ERC721 token on wallet portfolio

## Issues
- Cannot add/remove ERC721
    - ~Currently fetching owned tokens by rpc call `getPastLogs` to read block and filter by Transfer events~
    - ~But `getPastLogs` or `getPastEvents` only allows 2048 block in maximum, it will cause exception if too long without running the scan action. And waiting repeatedly fetching will also result in bad user experience~
    - because of immutable state to update
        - somehow the states became frozen under `mutations` with such encountering cases
            - `state.ercNftTokensCustom.push(...)`
            - `state.ercNftTokenIds.push(...)`
            - `Vue.set(state.walletBalance, ...)`
            - `state.ercNftTokenIds[index] = c`
            - `state.ercNftTokensCustom.splice(...)`
- Cannot display ERC721 image and metadata
    - `getTokenURI` has missing `/` to concat the IPFS ID hash
    - `ERCNftView`
        - getter `img` depends on `metadata` but didn't watch it
        - `img` didn't handle IPFS uri properly

## Solutions
- Workaround in the encountering cases, not really know why the states under `mutations` became frozen, 
    - Use `concat` instead of `push` to mutate array
    - Destruct the object and reassignment instead of `Vue.set` 
    - Slicing array to remove the target item instead of `arr.splice`
        - Ex: `[...slice(0, index), ...slice(index + 1)]`
- Use either CDN or manual uri to rewrite the `ipfs://` scheme
    - Extract the constant `CF_IPFS_BASE` to `constants.ts` file
    - Add `@Watch('metadata')` on getter `img` and add function `checkIPFSUri` to parse image url before rendering
- ~Refactor to fetch tokens from contract directly by extending the ERC721Enumerable from EIP standard, which allows to read tokens by address from contract directly. This is also current Avalanche implementation~
    - ~What modified~
        - ~Removed old `scanNewNfts` action~
        - ~Removed `ercNftTokenIds` in `ERCNftModuleState` because `tokenIds` will store in ERCNftToken instance~
        - ~Modify `canSupport` flag: both EIP `ERC721Metadata` and `ERC721Enumerable` should be extended~
        - ~Fetching owned tokens by combo of `balanceOf` and `tokenOfOwnerByIndex` directly from smart contract~

## Tested Cases
2 ERC721 contracts deployed @Kopernikus
    
    - 0x9cAad025f9e468d546fe7B80F788c0E09962217c
    - 0x1bF615d88B3cd4a15bdF79956d11d279Edff9AEb
- New wallet access (no related info stored in localStorage)
    - Add two collectibles to check if minted ERC721 tokens display image and metadata properly
    - Remove 1 added contract from `Add Collectible` modal to check if the view update properly
    - Re-add the removed contract to check if the minted tokens display properly

- Existing wallet access (reload the page and use the previous wallet to check recovering from localStorage works)
    - Reload page and re-login to check if previously added ERC721 tokens display properly
    - Remove 1 contract from `Add Collectible` modal to check if the view update properly
    - Add an ERC721 contract to check if the minted tokens display properly

---
### Update
Fix issue TECH-835

#### Root cause
The action `updateWalletBalance` do update the NFT token balance properly, but UI display all tokens without regarding the balance

#### How fixes
Filtering out the tokens with `0` balance in components,
- `wallet/portfolio/ERCNftFamilyRow.vue`
- `wallet/TopCards/BalanceCard/NftCol.vue`
